### ISSUE TITLE
Hibernate Reactive Session/EntityManager is closed

### ISSUE THREAD

**Issue by gwenneg** [quarkusio/quarkus#23269]
### Describe the bug

This may look similar to #22433 (which was fixed in `2.6.3.Final`) but this time the reproducer includes a resteasy-reactive request filter (see `RequestFilter` in the reproducer).

Here's the exception which is thrown randomly when the app is processing concurrent requests:

```
2022-01-28 15:11:21,411 ERROR [org.hib.rea.errors] (vert.x-eventloop-thread-4) HR000057: Failed to execute statement [$1select fruit0_.id as id1_0_, fruit0_.name as name2_0_ from known_fruits fruit0_ order by fruit0_.name]: $2could not execute query: java.util.concurrent.CompletionException: java.lang.IllegalStateException: Session/EntityManager is closed
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)
	at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1081)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2073)
	at io.vertx.core.Future.lambda$toCompletionStage$2(Future.java:360)
	at io.vertx.core.impl.future.FutureImpl$3.onSuccess(FutureImpl.java:141)
	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:60)
	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:211)
	at io.vertx.core.impl.future.PromiseImpl.tryComplete(PromiseImpl.java:23)
	at io.vertx.sqlclient.impl.QueryResultBuilder.tryComplete(QueryResultBuilder.java:102)
	at io.vertx.sqlclient.impl.QueryResultBuilder.tryComplete(QueryResultBuilder.java:35)
	at io.vertx.core.Promise.complete(Promise.java:66)
	at io.vertx.core.Promise.handle(Promise.java:51)
	at io.vertx.core.Promise.handle(Promise.java:29)
	at io.vertx.core.impl.future.FutureImpl$3.onSuccess(FutureImpl.java:141)
	at io.vertx.core.impl.future.FutureBase.lambda$emitSuccess$0(FutureBase.java:54)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:503)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.IllegalStateException: Session/EntityManager is closed
	at org.hibernate.internal.AbstractSharedSessionContract.checkOpen(AbstractSharedSessionContract.java:407)
	at org.hibernate.engine.spi.SharedSessionContractImplementor.checkOpen(SharedSessionContractImplementor.java:148)
	at org.hibernate.reactive.session.impl.ReactiveSessionImpl.checkOpen(ReactiveSessionImpl.java:1558)
	at org.hibernate.internal.AbstractSharedSessionContract.checkOpenOrWaitingForAutoClose(AbstractSharedSessionContract.java:413)
	at org.hibernate.internal.SessionImpl.getEntityUsingInterceptor(SessionImpl.java:603)
	at org.hibernate.loader.Loader.getRow(Loader.java:1610)
	at org.hibernate.loader.Loader.getRowFromResultSet(Loader.java:748)
	at org.hibernate.loader.Loader.getRowsFromResultSet(Loader.java:1047)
	at org.hibernate.reactive.loader.hql.impl.ReactiveQueryLoader.getRowsFromResultSet(ReactiveQueryLoader.java:223)
	at org.hibernate.reactive.loader.ReactiveLoaderBasedResultSetProcessor.reactiveExtractResults(ReactiveLoaderBasedResultSetProcessor.java:73)
	at org.hibernate.reactive.loader.hql.impl.ReactiveQueryLoader$1.reactiveExtractResults(ReactiveQueryLoader.java:72)
	at org.hibernate.reactive.loader.ReactiveLoader.reactiveProcessResultSet(ReactiveLoader.java:145)
	at org.hibernate.reactive.loader.ReactiveLoader.lambda$doReactiveQueryAndInitializeNonLazyCollections$0(ReactiveLoader.java:77)
	at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1072)
	... 21 more


2022-01-28 15:11:21,413 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (vert.x-eventloop-thread-4) HTTP Request to /fruits failed, error id: 51fc210d-cc94-487b-bfb9-5c5b593badfb-17: java.lang.IllegalStateException: Session/EntityManager is closed
	at org.hibernate.internal.AbstractSharedSessionContract.checkOpen(AbstractSharedSessionContract.java:407)
	at org.hibernate.engine.spi.SharedSessionContractImplementor.checkOpen(SharedSessionContractImplementor.java:148)
	at org.hibernate.reactive.session.impl.ReactiveSessionImpl.checkOpen(ReactiveSessionImpl.java:1558)
	at org.hibernate.internal.AbstractSharedSessionContract.checkOpenOrWaitingForAutoClose(AbstractSharedSessionContract.java:413)
	at org.hibernate.internal.SessionImpl.getEntityUsingInterceptor(SessionImpl.java:603)
	at org.hibernate.loader.Loader.getRow(Loader.java:1610)
	at org.hibernate.loader.Loader.getRowFromResultSet(Loader.java:748)
	at org.hibernate.loader.Loader.getRowsFromResultSet(Loader.java:1047)
	at org.hibernate.reactive.loader.hql.impl.ReactiveQueryLoader.getRowsFromResultSet(ReactiveQueryLoader.java:223)
	at org.hibernate.reactive.loader.ReactiveLoaderBasedResultSetProcessor.reactiveExtractResults(ReactiveLoaderBasedResultSetProcessor.java:73)
	at org.hibernate.reactive.loader.hql.impl.ReactiveQueryLoader$1.reactiveExtractResults(ReactiveQueryLoader.java:72)
	at org.hibernate.reactive.loader.ReactiveLoader.reactiveProcessResultSet(ReactiveLoader.java:145)
	at org.hibernate.reactive.loader.ReactiveLoader.lambda$doReactiveQueryAndInitializeNonLazyCollections$0(ReactiveLoader.java:77)
	at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1072)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2073)
	at io.vertx.core.Future.lambda$toCompletionStage$2(Future.java:360)
	at io.vertx.core.impl.future.FutureImpl$3.onSuccess(FutureImpl.java:141)
	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:60)
	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:211)
	at io.vertx.core.impl.future.PromiseImpl.tryComplete(PromiseImpl.java:23)
	at io.vertx.sqlclient.impl.QueryResultBuilder.tryComplete(QueryResultBuilder.java:102)
	at io.vertx.sqlclient.impl.QueryResultBuilder.tryComplete(QueryResultBuilder.java:35)
	at io.vertx.core.Promise.complete(Promise.java:66)
	at io.vertx.core.Promise.handle(Promise.java:51)
	at io.vertx.core.Promise.handle(Promise.java:29)
	at io.vertx.core.impl.future.FutureImpl$3.onSuccess(FutureImpl.java:141)
	at io.vertx.core.impl.future.FutureBase.lambda$emitSuccess$0(FutureBase.java:54)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:503)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)


2022-01-28 15:11:21,414 ERROR [org.jbo.res.rea.com.cor.AbstractResteasyReactiveContext] (vert.x-eventloop-thread-4) Request failed: java.lang.IllegalStateException: Session/EntityManager is closed
	at org.hibernate.internal.AbstractSharedSessionContract.checkOpen(AbstractSharedSessionContract.java:407)
	at org.hibernate.engine.spi.SharedSessionContractImplementor.checkOpen(SharedSessionContractImplementor.java:148)
	at org.hibernate.reactive.session.impl.ReactiveSessionImpl.checkOpen(ReactiveSessionImpl.java:1558)
	at org.hibernate.internal.AbstractSharedSessionContract.checkOpenOrWaitingForAutoClose(AbstractSharedSessionContract.java:413)
	at org.hibernate.internal.SessionImpl.getEntityUsingInterceptor(SessionImpl.java:603)
	at org.hibernate.loader.Loader.getRow(Loader.java:1610)
	at org.hibernate.loader.Loader.getRowFromResultSet(Loader.java:748)
	at org.hibernate.loader.Loader.getRowsFromResultSet(Loader.java:1047)
	at org.hibernate.reactive.loader.hql.impl.ReactiveQueryLoader.getRowsFromResultSet(ReactiveQueryLoader.java:223)
	at org.hibernate.reactive.loader.ReactiveLoaderBasedResultSetProcessor.reactiveExtractResults(ReactiveLoaderBasedResultSetProcessor.java:73)
	at org.hibernate.reactive.loader.hql.impl.ReactiveQueryLoader$1.reactiveExtractResults(ReactiveQueryLoader.java:72)
	at org.hibernate.reactive.loader.ReactiveLoader.reactiveProcessResultSet(ReactiveLoader.java:145)
	at org.hibernate.reactive.loader.ReactiveLoader.lambda$doReactiveQueryAndInitializeNonLazyCollections$0(ReactiveLoader.java:77)
	at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1072)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2073)
	at io.vertx.core.Future.lambda$toCompletionStage$2(Future.java:360)
	at io.vertx.core.impl.future.FutureImpl$3.onSuccess(FutureImpl.java:141)
	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:60)
	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:211)
	at io.vertx.core.impl.future.PromiseImpl.tryComplete(PromiseImpl.java:23)
	at io.vertx.sqlclient.impl.QueryResultBuilder.tryComplete(QueryResultBuilder.java:102)
	at io.vertx.sqlclient.impl.QueryResultBuilder.tryComplete(QueryResultBuilder.java:35)
	at io.vertx.core.Promise.complete(Promise.java:66)
	at io.vertx.core.Promise.handle(Promise.java:51)
	at io.vertx.core.Promise.handle(Promise.java:29)
	at io.vertx.core.impl.future.FutureImpl$3.onSuccess(FutureImpl.java:141)
	at io.vertx.core.impl.future.FutureBase.lambda$emitSuccess$0(FutureBase.java:54)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:503)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

### How to Reproduce?

Reproducer: https://github.com/gwenneg/quarkus-hibernate-reactive-server-request-filter

- Start the app with `./mvnw clean quarkus:dev`
- Run [jmeter.test.plan.zip](https://github.com/quarkusio/quarkus/files/7959301/jmeter.test.plan.zip) with JMeter
- The exception should show up in the log eventually

⚠️ If you change the Quarkus version to `2.3.1.Final`, the exception should never be thrown. As a consequence, this looks like a regression to us but I'll wait for a confirmation on that specific point before tagging the issue as a regression.

### Output of `uname -a` or `ver`

Linux glepage.remote.csb 4.18.0-348.2.1.el8_5.x86_64 #1 SMP Mon Nov 8 13:30:15 EST 2021 x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

openjdk version "11.0.13" 2021-10-19 LTS

### Quarkus version or git rev

2.6.3.Final and 2.7.0.Final

**Comment by gwenneg** [comment]
In case you have an issue with the enclosed zip file, here's the JMeter configuration I used:

![146910332-5e2d863d-f617-4310-b9ac-c84bcb78f581](https://user-images.githubusercontent.com/10584698/151563722-82037cd8-abe2-42d1-a7a6-d1137e7ebbbc.png)

![146910433-234300b3-88a2-4e1f-b436-2d3f207f139e](https://user-images.githubusercontent.com/10584698/151563730-0abf51ae-560b-4b19-9d72-6cdf819e133c.png)

**Comment by gwenneg** [comment]
@geoand I'm not sure whether this bug is caused by hibernate-reactive or resteasy-reactive as it's only happening when a request filter is involved, so I'm tagging you as well 😄

**Comment by geoand** [comment]
I don't really see how it could be caused by RESTEasy Reactive, but I'll try and take a look next week

**Comment by geoand** [comment]
I don't see anything wrong from a RESTEasy Reactive perspective.

Just for completeness sake I'll mention that the filter **does** run a Vert.x event-loop thread as expected.

**Comment by gwenneg** [comment]
Thanks for checking!

**Comment by geoand** [comment]
> However, the method annotated with @ServerRequestFilter (Uni<Response> filter(ContainerRequestContext requestContext) ) is returning an Uni which is being run interleaved with the Uni being returned by the method annotated with @Get

This should not be the case... When a `Uni` is being returned, then the request pipeline is suspended and only resumed when the result of the `Uni` comes back.

> @geoand / @FroMage ? It seems a bit ironical that RestEasy Reactive doesn't support a reactively implemented filter?

Not sure what you mean here. As I explained above, the reactive return type of the filter should result in suspending request (I will make sure there isn't some kind of bug here).

**Comment by Sanne** [comment]
> > However, the method annotated with @ServerRequestFilter (Uni filter(ContainerRequestContext requestContext) ) is returning an Uni which is being run interleaved with the Uni being returned by the method annotated with @get
> 
> This should not be the case... When a `Uni` is being returned, then the request pipeline is suspended and only resumed when the result of the `Uni` comes back.

It's clearly being interleaved in this case.

> 
> > @geoand / @FroMage ? It seems a bit ironical that RestEasy Reactive doesn't support a reactively implemented filter?
> 
> Not sure what you mean here. As I explained above, the reactive return type of the filter should result in suspending request (I will make sure there isn't some kind of bug here).

I'm referring to the points described on:
 https://github.com/quarkusio/quarkus/blob/65f0a3ddeb55c0b3b3c50df53c8994baf36bd71b/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/ServerRequestFilter.java#L53-L74

Most likely I'm just confused, but it seems to suggest that Uni responses should be used for blocking operations; it doesn't suggest what one is supposed to do in this case while having a "fully reactive" chain.

**Comment by geoand** [comment]
> It's clearly being interleaved in this case.

I'll take a closer look.

> Most likely I'm just confused, but it seems to suggest that Uni responses should be used for blocking operations; it doesn't suggest what one is supposed to do in this case while having a "fully reactive" chain.

I agree it's not phrased well. I'll fix it.

**Comment by Sanne** [comment]
thanks @geoand !

**Comment by geoand** [comment]
I am going to stand by my position that there is nothing wrong in the filter handling in RESTEasy Reactive.

I verified this by taking the reproducer and changing it in the following way:

```java
    @ServerRequestFilter
    public Uni<Response> filter(ContainerRequestContext requestContext) {
        return sf.withSession(s -> {
            requestContext.getHeaders().add("test", "true");
            return s.createNativeQuery("SELECT 1")
                    .getSingleResult()
                    .replaceWith(() -> null);
                }

        );
    }
```

Note that inside the `Uni` I am adding a test HTTP header.

Then I made changed the JAX-RS method to:

```java
    @GET
    public Uni<List<Fruit>> get(@HeaderParam("test") String testHeader) {
        if (!"true".equals(testHeader)) {
            throw new IllegalStateException("RR is severily misbehaving");
        }
        return sf.withTransaction((s,t) -> {
            return s.createNamedQuery("Fruits.findAll", Fruit.class)
                    .getResultList();
                }

        );
    }
```

Now when I hit this with a lot traffic, I do get 

```
2022-02-01 19:40:27,131 ERROR [org.hib.rea.errors] (vert.x-eventloop-thread-44) HR000057: Failed to execute statement [$1select fruit0_.id as id1_0_, fruit0_.name as name2_0_ from known_fruits fruit0_ order by fruit0_.name]: $2could not execute query: java.util.concurrent.CompletionException: java.lang.IllegalStateException: Session/EntityManager is closed
```

but nowhere do I see `RR is severily misbehaving` that should be thrown if the execution of the filter and resource method is interleaved.

What's more, I did the same experiment taking HR out of the picture entirely, and just used the test HTTP header I mentioned above, through a lot more load at it and it never once failed.

**Comment by Sanne** [comment]
I don't think you can test it in this way: the HR issue is only manifesting sporadically and with high load, and it's helped by the fact that the *span* of events being covered is quite long: we do need to send two operations to the RDBMS, and then wait for results for these.

But thanks for the idea, I will play with some variations of that. Worst case I will learn how wrong I am :)

**Comment by geoand** [comment]
I'm very interested in what you find!

Let me know so I can dig into this more if needed

**Comment by Sanne** [comment]
Complex discussion followed up:
 - https://quarkusio.zulipchat.com/#narrow/stream/187038-dev/topic/Vertx.20Local.20context

In short, I believe RestEasy Reactive should isolate the vertx context, but others aren't convinced of this; it might take some time.

In that case @gwenneg you won't be able to use context-bound objects, which are implicitly relied on if you use `withSession` or `withTransaction` ; as alternative you'll have to open sessions and transactions explicitly (e.g. `openSession()`), and `close()` them explicitly.

**Comment by Sanne** [comment]
Also: I've adapted the reproducer to verify our requirements without using Hibernate Reactive:
 - https://github.com/Sanne/quarkus-resteasy-reactive-vertx-context

It also fails only under load, but it might be a little simpler to understand why.

**Comment by geoand** [comment]
> It also fails only under load, but it might be a little simpler to understand why.

The reproducer is excellent in showing what the thorny issue is.

TL;DR: Hibernate Reactive expects `Vertx.currentContext().getLocal()` and `Vertx.currentContext().putLocal()` to work on a per-request basis. This does seem reasonable, but it appears like Hibernate Reactive was the only system that has this expectation.

**Comment by Sanne** [comment]
@gwenneg this was fixed by #23397

**Comment by gwenneg** [comment]
Thanks!

**Comment by mklueh** [comment]
> Complex discussion followed up:
> 
> * https://quarkusio.zulipchat.com/#narrow/stream/187038-dev/topic/Vertx.20Local.20context
> 
> In short, I believe RestEasy Reactive should isolate the vertx context, but others aren't convinced of this; it might take some time.
> 
> In that case @gwenneg you won't be able to use context-bound objects, which are implicitly relied on if you use `withSession` or `withTransaction` ; as alternative you'll have to open sessions and transactions explicitly (e.g. `openSession()`), and `close()` them explicitly.

Is this now not required anymore, as this issue is closed?
